juju/juju
# [JUJU-1745] Add support for the secret-remove hook

The secret-remove hook is implemented.
Plus support for deleting individual revisions rather than the whole secret.
Add a bug fix for secret deletion to clean up the owner label.

When a new secret revision is added, or a consumer tracks a new revision, we check to see if there are any revisions that are now unused. If so, we notify the unit agent and it can run the secret-remove hook. The hook can choose to delete that now obsolete revision if it wants. Once notified, it will not be notified again that the revision is obsolete.

## Checklist


## QA steps

Deploy a charm and add a secret with a label. I added a hook which logged the secret details in the secret-remove hook.
Grant access to another charm and get the secret content in that charm.
Add a new revision.
Do a secret get --update in the other charm.
The secret owner should see the secret remove hook called for revision 1.
Add a new revision. No hook.
Add a new revision - see a hook for the X-1 revision which is now unused.
Delete the obsolete revision.
From the CLI, use the secrets command to list revisions.
Delete the secret entirely.
Create a new one with the same label.
